server/datastore/mysql/software_installers.go

```diff
@@ -287,3 +323,98 @@ WHERE
 
 	return &dest, nil
 }
+
+func tmplNamedSQLCaseHostSoftwareInstallStatus(alias string) string {
+	return fmt.Sprintf(`
+	CASE WHEN %[1]s.post_install_script_exit_code IS NOT NULL
+		AND %[1]s.post_install_script_exit_code = 0 THEN
+		:installed
+	WHEN %[1]s.post_install_script_exit_code IS NOT NULL
+		AND %[1]s.post_install_script_exit_code != 0 THEN
+		:failed
+	WHEN %[1]s.install_script_exit_code IS NOT NULL
+		AND %[1]s.install_script_exit_code = 0 THEN
+		:installed
+	WHEN %[1]s.install_script_exit_code IS NOT NULL
+		AND %[1]s.install_script_exit_code != 0 THEN
+		:failed
+	WHEN %[1]s.pre_install_query_output IS NOT NULL
+		AND %[1]s.pre_install_query_output = '' THEN
+		:failed
+	WHEN %[1]s.host_id IS NOT NULL THEN
+		:pending
+	ELSE
+		NULL -- not installed from Fleet installer
+	END`, alias)
+}
+
+func (ds *Datastore) GetSummaryHostSoftwareInstalls(ctx context.Context, installerID uint) (*fleet.SoftwareInstallerStatusSummary, error) {
+	var dest fleet.SoftwareInstallerStatusSummary
+
+	stmt := fmt.Sprintf(`
+SELECT
+	COALESCE(SUM( IF(status = :pending, 1, 0)), 0) AS pending,
+	COALESCE(SUM( IF(status = :failed, 1, 0)), 0) AS failed,
+	COALESCE(SUM( IF(status = :installed, 1, 0)), 0) AS installed
+FROM (
+SELECT
+	software_installer_id,
+	%s AS status
+FROM
+	host_software_installs hsi
+WHERE
+	software_installer_id = :installer_id
+	AND id IN(
+		SELECT
+			max(id) -- ensure we use only the most recently created install attempt for each host
+			FROM host_software_installs
+		WHERE 
+			software_installer_id = :installer_id
+		GROUP BY
+			host_id)) s`, tmplNamedSQLCaseHostSoftwareInstallStatus("hsi"))
+
+	query, args, err := sqlx.Named(stmt, map[string]interface{}{
+		"installer_id": installerID,
+		"pending":      fleet.SoftwareInstallerPending,
+		"failed":       fleet.SoftwareInstallerFailed,
+		"installed":    fleet.SoftwareInstallerInstalled,
+	})
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "get summary host software installs: named query")
+	}
+
+	err = sqlx.GetContext(ctx, ds.reader(ctx), &dest, query, args...)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "get summary host software install status")
+	}
+
+	return &dest, nil
+}
+
+func (ds *Datastore) softwareInstallerJoin(installerID uint, status fleet.SoftwareInstallerStatus) (string, []interface{}, error) {
+	stmt := fmt.Sprintf(`JOIN (
+SELECT
+	host_id
+FROM
+	host_software_installs hsi
+WHERE
+	software_installer_id = :installer_id
+	AND hsi.id IN(
+		SELECT
+			max(id) -- ensure we use only the most recent install attempt for each host
```

Comment:
That's probably ok, but I _think_ depending on order of auto-generated ids is not recommended, they can have funky behaviors sometimes with transactions and skips and all... I used the `created_at` timestamp elsewhere for that (and `hsi.id = (SELECT ... LIMIT 1)` instead of `IN`, I figure it might help the db engine knowing that it will return only 1 row).

Reply:
@mna, I was on the fence about whether to go with the auto-generated id or the timestamp. My thinking for leaning toward the id was that it is the primary key that gets automatically incorporated into all other indices and I was hoping that might allow for some query optimization by the db engine, although that's just an uninformed hunch really. More research and investigation would be needed to see if that actually is actually a thing. If it works, I was hoping it might be a nifty way to simplify some of our timestamp comparisons.

I had considered the issue with skips and figured that wouldn't be a problem using `MAX` so long as the numbers are always increasing with each new request. Transactions are an interesting wrinkle though.

Reply:
I don't think there are performance concerns given that we already filter the sub-query by `host_id` and `software_installer_id`, which should use the index that we have for those 2 keys. Only a handful of rows would typically be checked for the latest one in that set.

That being said, I couldn't find any hard evidence that the auto-increment id should not be relied on in that way, so I'm fine with keeping it like that! My main concern is that if this fails, this could be tricky to identify as the issue.

Reply:
>  we already filter the sub-query by host_id and software_installer_id

Oops, actually that's what I do in my query but I see that here it's a bit more complex.

Reply:
Ok I ended up asking ChatGPT instead of trying to find the right article :D 

> In general, it's not recommended to rely solely on auto-increment IDs for ordering rows in insertion order in a MySQL database, especially as the database grows and under certain conditions. Here's why:
> 
> 1. **Concurrency**: In a highly concurrent environment, multiple transactions can be inserting rows simultaneously. While each row will still receive a unique auto-increment ID, the order in which these IDs are assigned might not reflect the exact insertion order due to concurrency control mechanisms like locks.
> 
> 2. **Transaction Rollback**: If a transaction is rolled back, any auto-increment IDs allocated within that transaction will be lost. This means that the insertion order based solely on auto-increment IDs may not reflect the actual order of successful insertions.
> 
> 3. **Partitioning and Sharding**: In distributed databases where data is partitioned or sharded across multiple nodes, maintaining a strict insertion order across all partitions or shards might not be feasible or efficient.
> 
> 4. **Optimizations and Maintenance**: MySQL might perform optimizations or maintenance tasks that could reorder rows physically on disk, breaking the insertion order.
> 
> Instead, if you need to retrieve rows in the order they were inserted, you should include a timestamp column that records the insertion time. You can then query the data using this timestamp column to achieve the desired ordering.
> 
> However, if your use case allows for relaxed ordering requirements, and you understand the limitations mentioned above, you can use auto-increment IDs as a rough approximation of insertion order. Just keep in mind that it might not always be accurate.

Reply:
> However, if your use case allows for relaxed ordering requirements, and you understand the limitations mentioned above, you can use auto-increment IDs as a rough approximation of insertion order. Just keep in mind that it might not always be accurate.

This is kind of where I ended up when deciding to try the id-based approach. For filtering and summary purposes, rough approximation with `MAX(id)` could be good enough. So the comment "ensure we use only the most recent install attempt for each host" is a bit of an overstatement. 

I'm definitely open to a timestamp-based approach. The problem I had was in formulating something that didn't add more complex row-by-row subqueries ordered by timestamp. WDYT?

Reply:
> The problem I had was in formulating something that didn't add more complex row-by-row subqueries ordered by timestamp.

Yeah, I agree that this is a more complex query than what I originally thought it was. You're right, we can certainly go on with that and address any concern/issue later on if needed. Maybe eventually having an additional field to indicate that this is the latest entry would help with those queries, at the expense of more rows to write on insert (update old ones and set new one as latest).